File: projects/rpn2tex/migrations/java-module-by-module/src/main/java/com/rpn2tex/Parser.java
package com.rpn2tex;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.List;
import java.util.Objects;

/**
 * Parses a stream of tokens into an Abstract Syntax Tree (AST) using a stack-based RPN algorithm.
 *
 * <p>The parser implements the standard Reverse Polish Notation (RPN) algorithm:
 * <ul>
 *   <li>Numbers are pushed onto a stack</li>
 *   <li>Operators pop two operands, create a BinaryOp node, and push the result</li>
 *   <li>At EOF, the stack must contain exactly one expression (the root of the AST)</li>
 * </ul>
 *
 * <p>Example usage:
 * <pre>{@code
 * List<Token> tokens = lexer.tokenize();
 * Parser parser = new Parser(tokens);
 * Expr ast = parser.parse();  // Returns the root of the AST
 * }</pre>
 *
 * <p>Error cases:
 * <ul>
 *   <li>Empty expression (no tokens before EOF)</li>
 *   <li>Insufficient operands for an operator (stack has fewer than 2 items)</li>
 *   <li>Too many operands (stack has more than 1 item at EOF)</li>
 * </ul>
 */
public final class Parser {
    private final List<Token> tokens;
    private int pos;

    /**
     * Constructs a new Parser with the given token stream.
     *
     * @param tokens the list of tokens to parse (must include EOF token at end)
     * @throws NullPointerException if tokens is null
     */
    public Parser(List<Token> tokens) {
        this.tokens = Objects.requireNonNull(tokens, "tokens must not be null");
        this.pos = 0;
    }

    /**
     * Parses the token stream into an AST.
     *
     * <p>This method implements the stack-based RPN parsing algorithm:
     * <ol>
     *   <li>Initialize an empty stack</li>
     *   <li>For each token:
     *     <ul>
     *       <li>NUMBER: create a Number node and push onto stack</li>
     *       <li>OPERATOR (+, -, *, /): pop right operand, pop left operand,
     *           create BinaryOp node, push onto stack</li>
     *       <li>EOF: stop processing</li>
     *     </ul>
     *   </li>
     *   <li>Validate that exactly one expression remains on the stack</li>
     * </ol>
     *
     * @return the root expression of the AST
     * @throws RpnException if parsing fails due to:
     *         <ul>
     *           <li>Empty expression</li>
     *           <li>Insufficient operands for an operator</li>
     *           <li>Too many operands remaining</li>
     *         </ul>
     */
    public Expr parse() throws RpnException {
        Deque<Expr> stack = new ArrayDeque<>();

        while (!isAtEnd()) {
            Token token = current();

            switch (token.type()) {
                case NUMBER -> {
                    Expr node = new Number(token.line(), token.column(), token.value());
                    stack.push(node);
                    advance();
                }
                case PLUS, MINUS, MULT, DIV -> {
                    if (stack.size() < 2) {
                        throw new RpnException(
                            String.format("Not enough operands for operator '%s'", token.value()),
                            token.line(),
                            token.column()
                        );
                    }
                    // Pop right operand first (LIFO order)
                    Expr right = stack.pop();
                    // Then pop left operand
                    Expr left = stack.pop();

                    // Map TokenType to operator string
                    String operator = switch (token.type()) {
                        case PLUS -> "+";
                        case MINUS -> "-";
                        case MULT -> "*";
                        case DIV -> "/";
                        default -> throw new IllegalStateException("Unexpected operator: " + token.type());
                    };

                    Expr node = new BinaryOp(token.line(), token.column(), operator, left, right);
                    stack.push(node);
                    advance();
                }
                case EOF -> {
                    // Stop processing at EOF
                    break;
                }
                default -> throw new RpnException(
                    String.format("Unexpected token type: %s", token.type()),
                    token.line(),
                    token.column()
                );
            }

            // Check if we hit EOF
            if (token.type() == TokenType.EOF) {
                break;
            }
        }

        // Validate final stack state
        if (stack.isEmpty()) {
            // Empty expression - no tokens before EOF
            Token eofToken = tokens.isEmpty() ?
                new Token(TokenType.EOF, "", 1, 1) :
                tokens.get(tokens.size() - 1);
            throw new RpnException("Empty expression", eofToken.line(), eofToken.column());
        }

        if (stack.size() > 1) {
            // Too many operands - more than one value remains on stack
            Token lastToken = pos > 0 ? tokens.get(pos - 1) : tokens.get(0);
            throw new RpnException(
                String.format("%d values remain on stack", stack.size()),
                lastToken.line(),
                lastToken.column()
            );
        }

        return stack.pop();
    }

    /**
     * Returns the current token without advancing.
     *
     * @return the current token
     */
    private Token current() {
        return tokens.get(pos);
    }

    /**
     * Advances to the next token.
     */
    private void advance() {
        if (!isAtEnd()) {
            pos++;
        }
    }

    /**
     * Checks if we've reached the end of the token stream.
     *
     * @return true if at end, false otherwise
     */
    private boolean isAtEnd() {
        return pos >= tokens.size() || current().type() == TokenType.EOF;
    }
}

File: projects/rpn2tex/migrations/go-module-by-module-1/parser.go
package main

import "fmt"

// ParserError represents a parsing error with the token where the error occurred.
type ParserError struct {
	Message string
	Token   Token
}

// Error implements the error interface for ParserError.
func (e *ParserError) Error() string {
	return fmt.Sprintf("%s at line %d, column %d",
		e.Message, e.Token.Line, e.Token.Column)
}

// Parser parses a sequence of tokens into an abstract syntax tree using
// stack-based evaluation for Reverse Polish Notation (RPN).
type Parser struct {
	Tokens []Token
	Pos    int
}

// NewParser creates a new Parser with the given token list.
func NewParser(tokens []Token) *Parser {
	return &Parser{
		Tokens: tokens,
		Pos:    0,
	}
}

// Parse parses the token stream into an AST expression tree.
// It implements stack-based RPN parsing:
// - NUMBER tokens are pushed as Number nodes onto the stack
// - OPERATOR tokens pop two operands, create a BinaryOp, and push the result
// - At EOF, the stack must contain exactly one expression
func (p *Parser) Parse() (Expr, error) {
	stack := []Expr{}

	for !p.atEnd() {
		token := p.current()
		p.advance()

		switch token.Type {
		case NUMBER:
			// Push number node onto stack
			node := &Number{
				Line:   token.Line,
				Column: token.Column,
				Value:  token.Value,
			}
			stack = append(stack, node)

		case PLUS, MINUS, MULT, DIV:
			// Pop two operands, create binary operation, push result
			if len(stack) < 2 {
				return nil, &ParserError{
					Message: fmt.Sprintf("Not enough operands for operator '%s'", token.Value),
					Token:   token,
				}
			}

			// Pop right operand (top of stack)
			right := stack[len(stack)-1]
			stack = stack[:len(stack)-1]

			// Pop left operand (new top of stack)
			left := stack[len(stack)-1]
			stack = stack[:len(stack)-1]

			// Map token type to operator string
			operator := p.tokenTypeToOperator(token.Type)

			// Create binary operation node
			node := &BinaryOp{
				Line:     token.Line,
				Column:   token.Column,
				Operator: operator,
				Left:     left,
				Right:    right,
			}
			stack = append(stack, node)

		case EOF:
			// Stop processing at EOF
			break

		default:
			return nil, &ParserError{
				Message: fmt.Sprintf("Unexpected token type: %s", token.Type),
				Token:   token,
			}
		}
	}

	// Validate the final stack
	if len(stack) == 0 {
		// Use the EOF token for error reporting
		eofToken := p.Tokens[len(p.Tokens)-1]
		return nil, &ParserError{
			Message: "Empty expression",
			Token:   eofToken,
		}
	}

	if len(stack) > 1 {
		// Multiple items left means missing operators
		// Use the position of the second-to-last token
		lastToken := p.Tokens[len(p.Tokens)-2]
		return nil, &ParserError{
			Message: fmt.Sprintf("Too many operands (missing operators): %d items left on stack", len(stack)),
			Token:   lastToken,
		}
	}

	return stack[0], nil
}

// current returns the current token without advancing.
func (p *Parser) current() Token {
	if p.Pos >= len(p.Tokens) {
		// Return EOF token if past the end
		return Token{Type: EOF, Value: "", Line: 0, Column: 0}
	}
	return p.Tokens[p.Pos]
}

// atEnd checks if the parser is at the end of the token stream.
func (p *Parser) atEnd() bool {
	return p.Pos >= len(p.Tokens) || p.current().Type == EOF
}

// advance moves to the next token and returns the previous current token.
func (p *Parser) advance() Token {
	token := p.current()
	if p.Pos < len(p.Tokens) {
		p.Pos++
	}
	return token
}

// tokenTypeToOperator converts a TokenType to its operator string.
func (p *Parser) tokenTypeToOperator(tt TokenType) string {
	switch tt {
	case PLUS:
		return "+"
	case MINUS:
		return "-"
	case MULT:
		return "*"
	case DIV:
		return "/"
	default:
		return ""
	}
}
